src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -1556,106 +1523,284 @@ internal static string CheckSuspiciousContent(Ast scriptBlockAst)
+                else if (upTo > upToMax) upTo = upToMax;
 
-        // Regular string signatures that can be detected with just string comparison.
-        private static HashSet<string> s_signatures = new HashSet<string>(StringComparer.OrdinalIgnoreCase) {
-            // Calling Add-Type
-            "Add-Type", "DllImport",
+                for (var i = 0; i < upTo; i++)
+                {
+                    var result = LookupHash(runningHash[i]);
+                    if (result != null) return result;
+                }
 
-            // Doing dynamic assembly building / method indirection
-            "DefineDynamicAssembly", "DefineDynamicModule", "DefineType", "DefineConstructor", "CreateType",
-            "DefineLiteral", "DefineEnum", "DefineField", "ILGenerator", "Emit", "UnverifiableCodeAttribute",
-            "DefinePInvokeMethod", "GetTypes", "GetAssemblies", "Methods", "Properties",
+                return null;
+            }
 
-            // Suspicious methods / properties on "Type"
-            "GetConstructor", "GetConstructors", "GetDefaultMembers", "GetEvent", "GetEvents", "GetField",
-            "GetFields", "GetInterface", "GetInterfaceMap", "GetInterfaces", "GetMember", "GetMembers",
-            "GetMethod", "GetMethods", "GetNestedType", "GetNestedTypes", "GetProperties", "GetProperty",
-            "InvokeMember", "MakeArrayType", "MakeByRefType", "MakeGenericType", "MakePointerType",
-            "DeclaringMethod", "DeclaringType", "ReflectedType", "TypeHandle", "TypeInitializer",
-            "UnderlyingSystemType",
+            /// <summary>
+            /// Scan a string for suspicious content.
+            ///
+            /// This is based on the Rubin-Karp algorithm, but heavily
+            /// modified to support searching for multiple patterns at
+            /// the same time.
+            ///
+            /// The key difference from Rubin-Karp is that we don't undo
+            /// the hash of the first character as we shift along in the
+            /// input.
+            ///
+            /// Instead, we can rely on knowing we need the hashes for
+            /// shorter strings anyway, so we reuse their values in
+            /// computing the hash for the longer patterns. This lets us
+            /// use a much simpler hash as well - we can avoid the use of
+            /// mod.
+            /// </summary>
+            /// <returns>The string matching the hash, or null.</returns>
+            public static string Match(string text)
+            {
+                // The longest pattern is 29 characters.
+                // The values in the array are the computed hashes of length
+                // index-1 (so runningHash[0] holds the hash for length 1).
+                var runningHash = new uint[29];
+
+                int longestPossiblePattern = 0;
+                for (int i = 0; i < text.Length; i++)
+                {
+                    uint h = text[i];
```

Comment:
Please change 'h' variable to 'c' to be consistent with HasNewPattern which uses 'c' as character variable.

Reply:
Consistency is over rated. Here `h` is not a character, but the hash value.

Reply:
uint h = text[i];    'h' is not a character from the text to be tested?  I though the hashes were stored in 'runningHash'?

Reply:
Yes, it's both, but I don't need 2 local variables.

Reply:
And hashes are stored in `runningHash`, but `h` is being used to compute the hash. One could say it is partially hashed. In previous iterations of the code, I had intermediate steps, so there were multiple assignments to `h` as part of hashing.